File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/Expressions/BracketedExpression.cs
using System;
using System.Collections.Generic;
using Lexy.Poc.Core.Parser;
using Lexy.Poc.Core.Parser.Tokens;

namespace Lexy.Poc.Core.Language.Expressions
{
    public class BracketedExpression : Expression
    {
        public string FunctionName { get; }
        public Expression Expression { get; }

        private BracketedExpression(string functionName, Expression expression, ExpressionSource source) : base(source)
        {
            FunctionName = functionName;
            Expression = expression;
        }

        public static ParseExpressionResult Parse(ExpressionSource source)
        {
            var tokens = source.Tokens;
            if (!IsValid(tokens))
            {
                return ParseExpressionResult.Invalid<BracketedExpression>("Not valid.");
            }

            var matchingClosingParenthesis = FindMatchingClosingBracket(tokens);
            if (matchingClosingParenthesis == -1)
            {
                return ParseExpressionResult.Invalid<BracketedExpression>("No closing bracket found.");
            }

            var functionName = tokens.TokenValue(0);
            var innerExpressionTokens = tokens.TokensRange(2, matchingClosingParenthesis - 1);
            var parseResult = ExpressionFactory.Parse(innerExpressionTokens, source.Line);

            var innerExpression = parseResult;

            var expression = new BracketedExpression(functionName, innerExpression, source);
            return ParseExpressionResult.Success(expression);
        }

        public static bool IsValid(TokenList tokens)
        {
            return tokens.Length > 1
                && tokens.IsTokenType<StringLiteralToken>(0)
                && tokens.OperatorToken(1, OperatorType.OpenBrackets);
        }

        internal static int FindMatchingClosingBracket(TokenList tokens)
        {
            if (tokens == null) throw new ArgumentNullException(nameof(tokens));

            var count = 0;
            for (var index = 0; index < tokens.Length; index++)
            {
                var token = tokens[index];
                if (!(token is OperatorToken operatorToken)) continue;

                if (operatorToken.Type == OperatorType.OpenBrackets)
                {
                    count++;
                }
                else if (operatorToken.Type == OperatorType.CloseBrackets)
                {
                    count--;
                    if (count == 0)
                    {
                        return index;
                    }
                }
            }

            return -1;
        }

        protected override IEnumerable<INode> GetChildren()
        {
            yield return Expression;
        }

        protected override void Validate(IParserContext context)
        {
        }
    }
}

File: typescipt/src/language/expressions/bracketedExpression.ts


export class BracketedExpression extends Expression {
   public string FunctionName
   public Expression Expression

   private BracketedExpression(string functionName, Expression expression,
     ExpressionSource source, SourceReference reference) : base(source, reference) {
     FunctionName = functionName;
     Expression = expression;
   }

   public static parse(source: ExpressionSource): ParseExpressionResult {
     let tokens = source.Tokens;
     if (!IsValid(tokens)) return ParseExpressionResult.Invalid<BracketedExpression>(`Not valid.`);

     let matchingClosingParenthesis = FindMatchingClosingBracket(tokens);
     if (matchingClosingParenthesis == -1)
       return ParseExpressionResult.Invalid<BracketedExpression>(`No closing bracket found.`);

     let functionName = tokens.TokenValue(0);
     let innerExpressionTokens = tokens.TokensRange(2, matchingClosingParenthesis - 1);
     let innerExpression = ExpressionFactory.Parse(innerExpressionTokens, source.Line);
     if (!innerExpression.IsSuccess) return innerExpression;

     let reference = source.CreateReference();

     let expression = new BracketedExpression(functionName, innerExpression.Result, source, reference);
     return ParseExpressionResult.Success(expression);
   }

   public static isValid(tokens: TokenList): boolean {
     return tokens.Length > 1
        && tokens.IsTokenType<StringLiteralToken>(0)
        && tokens.OperatorToken(1, OperatorType.OpenBrackets);
   }

   private static findMatchingClosingBracket(tokens: TokenList): number {
     if (tokens == null) throw new Error(nameof(tokens));

     let count = 0;
     for (let index = 0; index < tokens.Length; index++) {
       let token = tokens[index];
       if (!(token is OperatorToken operatorToken)) continue;

       if (operatorToken.Type == OperatorType.OpenBrackets) {
         count++;
       }
       else if (operatorToken.Type == OperatorType.CloseBrackets) {
         count--;
         if (count == 0) return index;
       }
     }

     return -1;
   }

   public override getChildren(): Array<INode> {
     yield return Expression;
   }

   protected override validate(context: IValidationContext): void {
   }

   public override deriveType(context: IValidationContext): VariableType {
     return Expression.DeriveType(context);
   }
}
